### ISSUE TITLE
你好，我模仿cosy2的pretrained_to_huggingface.py去写cosy3，能保存下来，但是后续trt run起来好像有问题

### ISSUE THREAD

**Issue by lzcchl** [nvidia-china-sae/mair-hub#35]
我看到cosyvoice项目里面有你的写trtllm文件夹，是cosyvoice2的，最近cosyvoice3出来了，想仿照你的把3的弄出来。

首先就是把cosyvoice3模型转成huggingface格式的，参考的是本项目中的rl-tutorial/cosyvoice_llm/pretrained_to_huggingface.py文件

但是v3貌似有一些变化，比如
1. 没有model.llm.llm_embedding，直接注释
2. llm_decoder.weight的维度是6761，v2版本好像是6564，于是我改了pad_to_multiple_of为256，new_lm_head.weight那里+200是为了能跑通
3. llm_decoder.bias为None，直接就不存了，最后的llm.config.use_bias改成False

于是我的代码如下，就是在你的基础上改的能保存下来，但是后续test_llm.py时貌似就有问题，您先看看v3的pretrained_to_huggingface.py有什么问题存在吗？

    import sys
    sys.path.insert(0, '/opt/code/CosyVoice')
    sys.path.insert(0, '/opt/code/CosyVoice/third_party/Matcha-TTS')

    from cosyvoice.cli.cosyvoice import CosyVoice2, AutoModel
    from argparse import ArgumentParser
    from transformers import AutoTokenizer, AutoModelForCausalLM
    import torch

    def get_args():
        parser = ArgumentParser()

        parser.add_argument(
            "--pretrained-cosyvoice3-path",
            type=str,
            default="/weights/Fun-CosyVoice3-0.5B-2512",
            help="Token2Wav path, default to %(default)r",
        )
        parser.add_argument(
            "--save-path",
            type=str,
            default='/weights/transformers_cosyvoice3_llm',
            help="The path to save the model",
        )
        args = parser.parse_args()
        return args

    if __name__ == "__main__":
        args = get_args()
        cosy3_model = AutoModel(
            model_dir=args.pretrained_cosyvoice3_path, load_vllm=False, load_trt=False, fp16=False
        )

        llm = cosy3_model.model.llm.llm.model

        speech_embedding = cosy3_model.model.llm.speech_embedding
        llm_decoder = cosy3_model.model.llm.llm_decoder
        # 没有这玩意了
        # llm_embedding = cosy3_model.model.llm.llm_embedding 

        tokenizer = AutoTokenizer.from_pretrained(f"{args.pretrained_cosyvoice3_path}/CosyVoice-BlankEN")
        special_tokens = {
            'eos_token': '<|endoftext|>',
            'pad_token': '<|endoftext|>',
            'additional_special_tokens': [
                '<|im_start|>', '<|im_end|>', '<|endofprompt|>',
                '[breath]', '<strong>', '</strong>', '[noise]',
                '[laughter]', '[cough]', '[clucking]', '[accent]',
                '[quick_breath]',
                "<laughter>", "</laughter>",
                "[hissing]", "[sigh]", "[vocalized-noise]",
                "[lipsmack]", "[mn]"
            ]
        }
        tokenizer.add_special_tokens(special_tokens)

        # original_tokenizer_vocab_size=151663
        original_tokenizer_vocab_size = len(tokenizer)
        # https://www.modelscope.cn/models/FunAudioLLM/Fun-CosyVoice3-0.5B-2512/file/view/master/cosyvoice3.yaml?status=1#L26
        cosyvoice_token_size = 6561
        new_tokens = [f"<|s_{i}|>" for i in range(cosyvoice_token_size)] + [
            "<|eos1|>", "<|eos2|>", "<|eos3|>", "<|sos|>", "<|task_id|>"
        ]
        num_added_tokens = tokenizer.add_tokens(new_tokens)

        llm.resize_token_embeddings(len(tokenizer), pad_to_multiple_of=256)
        vocab_size = llm.get_input_embeddings().weight.shape[0]

        feature_size = speech_embedding.embedding_dim
        new_lm_head = torch.nn.Linear(in_features=feature_size, out_features=vocab_size, bias=False)

        with torch.no_grad():
            # set the weight and bias of the new lm_head to 0
            new_lm_head.weight.data.zero_()
            # new_lm_head.bias.data.zero_()
            # 甭+3了，因为llm_decoder维度在cosy3变了, +200是为了维度一致...至于为什么俺也不知, 而bias整个为None了
            new_lm_head.weight[original_tokenizer_vocab_size:original_tokenizer_vocab_size+cosyvoice_token_size+200] = llm_decoder.weight
            # new_lm_head.bias[original_tokenizer_vocab_size:original_tokenizer_vocab_size+cosyvoice_token_size] = llm_decoder.bias

        llm.lm_head = new_lm_head
        input_embeddings = llm.get_input_embeddings()

        with torch.no_grad():
            input_embeddings.weight[original_tokenizer_vocab_size:original_tokenizer_vocab_size+cosyvoice_token_size+200] = speech_embedding.weight
            # input_embeddings.weight[original_tokenizer_vocab_size+cosyvoice_token_size+3:original_tokenizer_vocab_size+cosyvoice_token_size+3+2] = llm_embedding.weight

        eos_token_ids = [original_tokenizer_vocab_size + cosyvoice_token_size, original_tokenizer_vocab_size + cosyvoice_token_size + 1, original_tokenizer_vocab_size + cosyvoice_token_size + 2]
        llm.generation_config.eos_token_id = eos_token_ids
        llm.generation_config.temperature = 1.0
        llm.generation_config.top_p = 0.8
        llm.generation_config.top_k = 25

        llm.config.eos_token_id = original_tokenizer_vocab_size + cosyvoice_token_size
        llm.config.vocab_size = vocab_size
        llm.config.tie_word_embeddings = False
        llm.config.use_bias = False
        llm.to(torch.bfloat16)
        llm.save_pretrained(args.save_path)

        TEMPLATE = "{%- for message in messages %}{%- if message['role'] == 'user' %}{{- '<|sos|>' + message['content'] + '<|task_id|>' }}{%- elif message['role'] == 'assistant' %}{{- message['content']}}{%- endif %}{%- endfor %}"
        tokenizer.chat_template = TEMPLATE
        tokenizer.save_pretrained(args.save_path)

**Comment by yuekaizhang** [comment]
@lzcchl 你好，可以参考这个仓库 https://github.com/Brakanier/FastCosyVoice/blob/main/scripts/convert_cosyvoice3_to_hf.py

实现好的话，也欢迎 pr ！

**Comment by lzcchl** [comment]
@yuekaizhang 感谢回复，我使用这个俄罗斯老哥的代码生成hf模型，然后再使用 
https://github.com/nvidia-china-sae/mair-hub/blob/main/rl-tutorial/cosyvoice_llm/infer.py 来运行，结果还是不对，outputs = model.generate这里出来的结果感觉就短了三分之二。

infer.py 里面我相应的修改了
1. TEMPLATE
2. codec_decoder.frontend.frontend_zero_shot 和 codec_decoder.model.hift.inference 参数改成当初cosy3版本的参数
3. token2wav_model使用v3模型

我先在那边问问吧，这边等我和那边讨论。
https://github.com/Brakanier/FastCosyVoice/issues/15

**Comment by lzcchl** [comment]
@yuekaizhang 发现问题所在了，貌似是sos, task_id的的问题，v3中sos=6561, eos_token=6562, task_id=6563, fill_token=6564，和v2可能有差异，现在能得到正确的结果

提供一下现在的能跑对的代码：

    import sys
    sys.path.insert(0, '/opt/code/CosyVoice')
    sys.path.insert(0, '/opt/code/CosyVoice/third_party/Matcha-TTS')

    from transformers import AutoTokenizer, AutoModelForCausalLM
    import torch
    import soundfile as sf
    from cosyvoice.cli.cosyvoice import AutoModel
    from argparse import ArgumentParser
    import numpy as np
    import s3tokenizer
    import soundfile as sf

    ORIGINAL_VOCAB_SIZE = 151924


    def get_args():
        parser = ArgumentParser()

        parser.add_argument(
            "--token2wav-path",
            type=str,
            default='/weights/Fun-CosyVoice3-0.5B-2512',
            help="Token2Wav path, default to %(default)r",
        )
        parser.add_argument(
            "--prompt-text",
            type=str,
            default="吃燕窝就选燕之屋，本节目由26年专注高品质燕窝的燕之屋冠名播出。豆奶牛奶换着喝，营养更均衡，本节目由豆本豆豆奶特约播出。",
            help="The prompt text",
        )
        parser.add_argument(
            "--prompt-speech-path",
            type=str,
            default="/voice_file/prompt_audio.wav",
            help="The path to the prompt speech",
        )
        parser.add_argument(
            "--input-text",
            type=str,
            default='身临其境，换新体验。塑造开源语音合成新范式，让智能语音更自然。',
            help="The input text",
        )
        parser.add_argument(
            "--model-path",
            type=str,
            default='/weights/transformers_cosyvoice3_llm',
            help="The path to the model",
        )
        parser.add_argument(
            "--speech_tokenizer_model_path",
            type=str,
            default='/weights/Fun-CosyVoice3-0.5B-2512/speech_tokenizer_v3.onnx',
            help="path to speech_tokenizer_v3.onnx",
        )
        
        args = parser.parse_args()
        return args


    def audio_decode_cosyvoice(audio_tokens, tts_text, prompt_text, prompt_speech_path, codec_decoder):

        # 函数的参数要与当前版本的cosy代码对上
        model_inputs_dict = codec_decoder.frontend.frontend_zero_shot(
            tts_text, prompt_text, prompt_speech_path, 24000, ''
        )
        tts_mel, _ = codec_decoder.model.flow.inference(
            token=audio_tokens.to(codec_decoder.model.device),
            token_len=torch.tensor([audio_tokens.shape[1]], dtype=torch.int32).to(codec_decoder.model.device),
            prompt_token=model_inputs_dict["flow_prompt_speech_token"].to( codec_decoder.model.device),
            prompt_token_len=model_inputs_dict["flow_prompt_speech_token_len"].to(codec_decoder.model.device),
            prompt_feat=model_inputs_dict["prompt_speech_feat"].to(codec_decoder.model.device),
            prompt_feat_len=model_inputs_dict["prompt_speech_feat_len"].to(codec_decoder.model.device),
            embedding=model_inputs_dict["flow_embedding"].to(codec_decoder.model.device),
            finalize=True,
            streaming=False,
        )

        # 这里v2与v3不一样了
        audio_hat, _ = codec_decoder.model.hift.inference(
            # speech_feat=tts_mel, cache_source=torch.zeros(1, 1, 0)
            speech_feat=tts_mel, finalize=True
        )

        return audio_hat


    def extract_speech_ids(speech_tokens_str):
    
        speech_ids = []
        for token_str in speech_tokens_str:
            if token_str.startswith('<|s_') and token_str.endswith('|>'):
                num_str = token_str[4:-2]

                num = int(num_str)
                speech_ids.append(num)
            else:
                print(f"Unexpected token: {token_str}")
        return speech_ids


    if __name__ == '__main__':
        args = get_args()
        device = torch.device("cuda")

        tokenizer = AutoTokenizer.from_pretrained(args.model_path)
        model = AutoModelForCausalLM.from_pretrained(args.model_path)
        model.eval() 
        model.to(device)

        token2wav_model = AutoModel(
            model_dir=args.token2wav_path, load_trt=False, fp16=False
        )

        audio_tokenizer = s3tokenizer.load_model(args.speech_tokenizer_model_path).to(device)

        # audio_tokenizer
        waveform, sr = sf.read(args.prompt_speech_path)

        mels = []
        wav_array = torch.from_numpy(np.array(waveform, dtype=np.float32)).to(device)
        wav_len = len(wav_array)
        wav = wav_array.squeeze(0)
        mels.append(s3tokenizer.log_mel_spectrogram(wav))

        mels, mels_lens = s3tokenizer.padding(mels)
        codes, codes_lens = audio_tokenizer.quantize(mels.to(device), mels_lens.to(device))
        codes = codes.clone()# + ORIGINAL_VOCAB_SIZE
        prompt_speech_tokens = codes[0, :codes_lens[0].item()]
        prompt_speech_tokens = prompt_speech_tokens.cpu().numpy().tolist()
        prompt_speech_str = ''.join([f'<|s_{t}|>' for t in prompt_speech_tokens])

        with torch.no_grad():
            # # Tokenize the text
            # chat = [
            #     {"role": "user", "content": f"{args.input_text}"},
            #     {"role": "assistant", "content": ""}
            # ]
            # if 'system' in tokenizer.chat_template:
            #     tokenizer.chat_template = TEMPLATE
            # input_ids = tokenizer.apply_chat_template(
            #     chat, 
            #     tokenize=True, 
            #     return_tensors='pt', 
            #     continue_final_message=True
            # )
            # input_ids = input_ids.to(device)

            # https://github.com/Brakanier/FastCosyVoice/blob/main/fastcosyvoice/cosyvoice.py#L1046
            # <|s_6561|> = sos, <|s_6563|> = task_id
            prompt = f"<|s_6561|>{args.prompt_text + args.input_text}<|s_6563|>{prompt_speech_str}"
            input_ids = tokenizer.encode(prompt, return_tensors='pt').to(device)

            # Generate the speech autoregressively
            outputs = model.generate(
                input_ids,
                max_length=2048,  # We trained our model with a max length of 2048
                do_sample=True,    # True False
                top_p=0.95,           #  Adjusts the diversity of generated content
                temperature=0.8,   #  Controls randomness in output,
                repetition_penalty=1.1,
                top_k=15,           # 设置25容易多出一段重复的话
            )
            # Extract the speech tokens
            generated_ids = outputs[0][input_ids.shape[1]:-1]
            print(generated_ids.shape)

            speech_tokens = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)  

            # Convert  token <|s_23456|> to int 23456 
            speech_tokens = extract_speech_ids(speech_tokens)

            speech_tokens = (torch.tensor(speech_tokens)).cuda().unsqueeze(0)

            audio_hat = audio_decode_cosyvoice(
                speech_tokens,
                args.input_text,
                args.prompt_text,
                args.prompt_speech_path,
                token2wav_model,
            )

            audio = audio_hat.squeeze(0).cpu().numpy()
            sf.write("gen.wav", audio, 24000)

**Comment by yuekaizhang** [comment]
@lzcchl  感谢，后续有实验结果也希望能分享一下！

**Comment by yuekaizhang** [comment]
@lzcchl  我试了一下，应该是输入要这样修改， https://github.com/FunAudioLLM/CosyVoice/blob/main/runtime/triton_trtllm/infer_cosyvoice3.py#L175
